# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadManager.cpp
# Commit: 8132dd939023
# Full Hash: 8132dd9390233328c8da7072f36859424a2b079d
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-10-14 21:42:48
# Regressor Bug: 1657404
# File Overlap Count: 2
# Description:
#   Bug 1657404 - Change PGamepadEventChannel to "refcounted protocol" r=handyman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D93119
# ==============================================================================

diff -r 186ca9e93e1d -r 8132dd939023 dom/gamepad/GamepadManager.cpp
--- a/dom/gamepad/GamepadManager.cpp	Wed Oct 07 11:34:02 2020 +0000
+++ b/dom/gamepad/GamepadManager.cpp	Wed Oct 14 15:58:36 2020 +0000
@@ -115,8 +115,9 @@
 }
 
 void GamepadManager::StopMonitoring() {
-  for (uint32_t i = 0; i < mChannelChildren.Length(); ++i) {
-    PGamepadEventChannelChild::Send__delete__(mChannelChildren[i]);
+  for (auto& channelChild : mChannelChildren) {
+    PGamepadEventChannelChild::Send__delete__(channelChild);
+    channelChild = nullptr;
   }
   if (gfx::VRManagerChild::IsCreated()) {
     gfx::VRManagerChild* vm = gfx::VRManagerChild::Get();
@@ -149,15 +150,12 @@
       return;
     }
 
-    GamepadEventChannelChild* child = new GamepadEventChannelChild();
-    PGamepadEventChannelChild* initedChild =
-        actor->SendPGamepadEventChannelConstructor(child);
-    if (NS_WARN_IF(!initedChild)) {
+    RefPtr<GamepadEventChannelChild> child(GamepadEventChannelChild::Create());
+    if (!actor->SendPGamepadEventChannelConstructor(child.get())) {
       // We are probably shutting down.
       return;
     }
 
-    MOZ_ASSERT(initedChild == child);
     mChannelChildren.AppendElement(child);
 
     if (gfx::VRManagerChild::IsCreated()) {